[minor] Add Upgrade FVT Pipeline and Tekton Build System

alequint opened the pull request:
## Overview

This PR contains upgrade-test pipeline as well as automation code to generate and release pipelineruns in artifactory (and some other misc fixes like yml x yaml standards). Other PRs related to this one:

- `ansible-devops` (**merge before this**):
  - Add skip_compatibility_check for upgrade error path test (https://github.com/ibm-mas/ansible-devops/pull/595), originally created by @liyongc
  - Fix timeout in OperatorCondition Checking (https://github.com/ibm-mas/ansible-devops/pull/631) created by @alequint 

- `upg-stable` (as reference only): here is the travis that will trigger upgrade-template: https://github.ibm.com/maximoappsuite/upg-stable/blob/master/.travis.yml. Here we have:
  - minimum variables needed to install, test and upgrade
  - the variables representing values related to upgrade process itself are defined using `source_` and `target_` prefixes. Source and Target of some of them are equal today like ocp_version, cpd_product_version, mas_catalog_version, but are prepared to be different as we evolve upgrade test and script itself. Today, the only source/target vars are the channels of core and applications, following phase 1 of upgrade test epic.
  - difference from regular fvts is that mas_tekton's run-pipeline is used to trigger upgrade-test pipeline, as it uses automated generated pipelunerun released in artifactory, as will be explained along the PR


## Content

- New `upgrade-test` pipeline, which is an integration of 3 other existing pipelines:
  - `install`, `fvt` (which is same as current fvt-core, but plan is expand it to contain fvt of all the applications) and `upgrade`
- Three new Tasks capable to run these three pipelines (in `tekton/task-templates`):
  - `fvt-run-pipeline`
  - `install-run-pipeline`
  - `upgrade-run-pipeline`
  - **Note**: these tasks run a role inside [mas_tekton](https://github.ibm.com/maximoappsuite/mas-tekton) called `run-pipeline`. This role simply download a pipelinerun **template** from artifactory (as released by cli) and apply it in the cluster. It is expected environment is filled with the right variables to be used to fill up pipelinerun template; these variables are set by the run-pipeline tasks previously mentioned (which come from upgrade-test pipeline's taskdefs)
- Added more automation to generate tekton. Considering we have now tasks capable to run pipelines, we want that taskdefs, tasks and pipelineruns being automatically generated. It means that we just need to add parameter in the proper param template file and don't worry with the rest, as the automation will organize it all (add it in the task, in pipeline, in taskdefs, in pipelinerun itself etc). Manual settings are excluded from automation by using a special manualsettings file, so they are excluded from the loop and , must be set manually (it happens when the value does not come from owner pipeline, but we know the value to use). Tekton tree is like this now:

```
> tekton
  > generate-tekton.yml (playbook)
  > pipeline-templates
    > params
      > common.yml.j2
      > upgrade-test.yml.j2
      > upgrade.yml.j2
      > (others already there)
    > upgrade-test.yml.j2
  > pipelinerun-templates
    > fvt.yml.j2
    > install-with-fvt.yml.j2
    > install.yml.j2
    > upgrade-test.yml.j2
    > upgrade.yml.j2
  > task-templates
  > pipelines
  > pipelineruns
  > tasks
```
- Build process has been updated to upload pipelineruns to artifactory same way ibm-mas-tekton.yml. run-pipeline role then will down load the pipelinerun to apply and will have, in their environment all the variables needed to update it.
- Standardize use of `yml` instead of `yaml`. We had a mix of these two extensions in cli; then I picked up `yml` as it seems the one mostly used around, including `ansible-devops` (though even there we have a mix of standards in some parts)
- **Important** to say that in this first drop I will include nothing beyond core tests in fvt pipeline. I want to be sure we keep running that upgrade test against mas core first and then start including the applications one by one.

## Test Evidences

Complete set of pipelines triggered by upgrade-test pipeline. This test has been made with **mas core** and **optimizer**.

<img width="1061" alt="image" src="https://user-images.githubusercontent.com/15021471/215496967-c04a3b46-6978-47dc-a663-4fe6a8e9bb5f.png">

Commit:
[patch] Trigger build

Commit:
[minor] Upgrade Test and Tekton Automation

Commit:
[minor] Upgrade Test and Tekton Automation

Commit:
[minor] Automation to generate pipelineruns

Commit:
[minor] Building pipelinerun

Commit:
[minor] Build and upload pipelinerun artifacts

Commit:
[patch] Fix pipelinerun build

Commit:
[patch] Fix pipelinerun build

Commit:
[patch] Fix

Commit:
[patch] Fix

Commit:
[minor] Added test pipelinerun

Commit:
[minor] Added minor pipelinerun

Commit:
[patch] Fix pipelinerun names

Commit:
[patch] More fixes

Commit:
[patch] More fixes

Commit:
[patch] Fixes

Commit:
[patch] Fix parameter content

Commit:
[patch] Fixes

Commit:
[patch] Fix gitignore

Commit:
[patch] Fix

Commit:
[patch] Fix run templates

Commit:
[patch] Task templates directory + misc fixes

Commit:
Merge branch 'master' of github.com:ibm-mas/cli into upgtest

Commit:
[minor] run pipeline taskdefs + misc fixes

Commit:
[minor] Review pipeline run uploads action

Commit:
[patch] Fix identation

Commit:
[patch] Fix upgrade-test parameter values

Commit:
[patch] Fix fvt task stepTemplates

Commit:
[patch] Fix fvt-run task parameters

Commit:
[patch] More fixes

Commit:
[patch] Add install + fvt parameters in fvt only pipeline

Commit:
[patch] Add ansible_fvt_version for pre release tests

Commit:
[minor] Optional suffix to pipelinerun name

Commit:
[patch] pipelinerun name suffix

Commit:
[patch] adding generated files

Commit:
[patch] Remove extra dash in pipelinerun names

alequint committed:
[patch] Add 2 more vars to support upgrade report (merge into upgtest branch) (#137)

Co-authored-by: liyongc <liyongc@cn.ibm.com>

Commit:
[patch] Adding prerease devops 12.6.1-pre.supportskipcompacheck

Commit:
[minor] Merge master

Commit:
Merge branch 'upgtest' of github.com:ibm-mas/cli into upgtest

Commit:
Review extensions and gitignore

Commit:
[patch] Fix build

Commit:
[patch] Test pre release ansible-devops

Commit:
[patch] Added skip_compatibility_check

Commit:
[patch] Refactor

Commit:
[patch] More fixes in tekton automation

Commit:
[patch] More fixes

Commit:
{patch] Fix

Commit:
[patch] Fixes

Commit:
[patch] Fixes again

Commit:
[patch] Fixes and Fixes

Commit:
[patch] Incorporating common parameters to deal with upgrade ones properly

Commit:
[patch] Fixes

Commit:
[patch] Fixes

Commit:
[patch] Patched ansible-devops

Commit:
[patch] Make sls channel and mas catalog source parameter as source/target

Commit:
[patch] Make sls channel and mas catalog source parameter as source/target

Commit:
[patch] Merge branch 'upgtest' of github.com:ibm-mas/cli into upgtest

Commit:
[patch] Remove generated files + update gitignore

Commit:
[patch] Create pipelineruns destination directory

Commit:
[patch] Rename to follow yml standard

Commit:
[patch] Merge branch 'master' of github.com:ibm-mas/cli into upgtest

alequint committed:
Merge branch 'master' into upgtest

Commit:
[patch] Enhance automation for run-pipeline tasks

Commit:
Merge branch 'upgtest' of github.com:ibm-mas/cli into upgtest

Commit:
Merge branch 'master' of github.com:ibm-mas/cli into upgtest

Commit:
[patch] Rename parameter

Commit:
[patch] Using pre-release of ansible-fvt

Commit:
[patch] Passing test parameters to all tasks

Commit:
[minor] Test Type and Test Phase

Commit:
[patch] Missed parameters

Commit:
[patch] New pre-release ansible-devops

Commit:
[patch] Master Conflicts + new devops

Commit:
[patch] Fix

Commit:
[patch] Make yaml into yml

Commit:
[patch] Replace ansible fvt by mas tekton

Commit:
[minor] Upload pipelienrun definitions to artifactory

Commit:
[patch] Compat ansible-fvt yaml rather than yml

Commit:
[patch] Rename

Commit:
[patch] Add cli_version common parameter

Commit:
[patch] Moved artifactory parameters to common params

Commit:
[patch] typogit status! :(

Commit:
[patch] add skip compatibility param in upgrade tasks

durera committed:
Merge branch 'master' into upgtest

durera commented on .github/workflows/build-cli.yml:
```diff
@@ -29,16 +29,27 @@ jobs:
-          name: ibm-mas-tekton-${{ env.VERSION }}.yaml
-          path: ${{ github.workspace }}/image/cli/mascli/templates/ibm-mas-tekton.yaml
+          name: ibm-mas-tekton-${{ env.VERSION }}.yml
+          path: ${{ github.workspace }}/image/cli/mascli/templates/ibm-mas-tekton.yml
```
Wouldn't change this, turns this into a breaking change.  we use `yml` inside ansible code, and `yaml` for things that a user will `oc apply -f` to align with what seems or seemed at the time, the generally accepted standard in both contexts.

alequint replied:
Done. All extensions revisited following our chat in slack:
- Keep `yml.j2` for all `j2` files
- Everything else `yaml` except by:
- generate-tekton playbook, kept as `yml` (as it is ansible)

durera commented on Makefile:
```diff
@@ -24,10 +24,10 @@ run:
 
 create:
-	oc apply -f tmp/deployment.yaml
+	oc apply -f tmp/deployment.yml
```
As above, yaml (openshift/kubernetes) versus yml (ansible), don't think we really want to change this.

alequint replied:
Done. See comment above

durera commented on tekton/pipeline-templates/params/common.yml.j2:
```diff
@@ -0,0 +1,81 @@
+# CLI Version
```
Bad merge possibly?  `cli_version` is no more after the last round updates to optimize development flow.

alequint replied:
Same reason of the "fvt" parameters described below... `cli_version` is used inside run-pipeline task to download the right version of pipelineruns from artifactory.

durera replied:
Makes sense ... Should move/set that param specifically for that task now, rather than have it in the global/shared param definition.

durera commented on tekton/pipeline-templates/params/common.yml.j2:
```diff
@@ -0,0 +1,81 @@
+  description: version of mas cli to be used in tasks and dependencies
+  default: ""
+
+# Registry
```
`common` to me implies there are params used in every pipeline, but these `fvt_` params would only be used in the fvt pipelines, not the "normal" ones ... am I misunderstanding the context of "common" here?

alequint replied:
Here is what happen @durera :

When upgrade-test pipeline runs other pipelines, it will be running, in the end, the run-pipeline role  (this one [here](https://github.ibm.com/maximoappsuite/mas-tekton/tree/master/ansible/ibm/mas_tekton/roles/run_pipeline))  inside the new mas-tekton container.

That role then, needs to donwload the pipelinerun from artifactory (that is why the artifactory key). Also, the task running inside that pipeline runs mas-tekton container, which lives in artifactory (that is why the registry). Considering we use the param list to automate also the different taskdefs params inside the upgrade-test pipeline and even the task's params and stepTemplate envs (and so pass that to run-pipeline) task, I thought would make sense to have those settings in all the pipelines (just to simplify automation, as these infos will not be used by the pipeline itself, but by the runpipeline task that will eventually run them). I am treating the pipeline and the correspondent run-pipeline task basically as the same thing parameterwise, as 95% of params are the same)

What I could do here is to rename those attributes to remove all these `fvt_` prefixes for the items in common to avoid confusion (and add some comments about why they are there). What do you think?

durera replied:
Having spent more time working on this PR today, I think the confusion comes from the naming of this ... it should really be another pair of `devops_xxx` variables, their purpose is all about defining the records in the database that are created/updated to record the results of the tekton actions.

durera commented on tekton/pipeline-templates/params/fvt.yml.j2:
```diff
@@ -1,13 +1,3 @@
-# FVT Image Container Registry
```
See previous comment, I'm confused why these FVT specific params are being moved into common, that doesn't make sense to me.

alequint replied:
See comment above

durera commented on tekton/pipeline-templates/params/helper/install-list.yml.j2:
```diff
@@ -0,0 +1,2 @@
+params:
```
This file seems like an unnecessary layer of abstraction compared to whatever is including this including this content directly .. but might make sense as I get further through the changes ...

alequint replied:
https://ibm-watson-iot.slack.com/archives/C0195MVCEUD/p1675098979911869?thread_ts=1675088015.554689&cid=C0195MVCEUD

durera commented on tekton/pipeline-templates/params/helper/upgrade-list.yml.j2:
```diff
@@ -0,0 +1,2 @@
+params:
```
This file seems like an unnecessary layer of abstraction compared to whatever is including this including this content directly .. but might make sense as I get further through the changes ...

alequint replied:
https://ibm-watson-iot.slack.com/archives/C0195MVCEUD/p1675098979911869?thread_ts=1675088015.554689&cid=C0195MVCEUD

durera commented on tekton/tasks/cert-manager.yml:
```diff
@@ -48,6 +56,11 @@ spec:
         value: $(params.devops_suite_name)
       - name: DEVOPS_BUILD_NUMBER
         value: $(params.devops_build_number)
+      # Optional parameters to categoriz result to be persisted
```
Small typo in various places to find/replace fix `categoriz` -> `categorize`

alequint replied:
Done

durera reviewed (changes requested):
See individual comments ... 

Also, local built version of `ibm-mas_devops.tar.gz` needs to be deleted before this can be merged.  If there's a branch of ansible-devops that needs to be merged before this put a link in the PR description, e.g. like in this PR https://github.com/ibm-mas/cli/pull/124, and we'll get that released before this goes to master

Commit:
[patch] Removed pre-release ansible-devops

Commit:
Merge branch 'master' of github.com:ibm-mas/cli into upgtest

Commit:
Merge branch 'upgtest' of github.com:ibm-mas/cli into upgtest

Commit:
[patch] Fixed categoriz typo

Commit:
[patch] Update change logs

Commit:
[patch] Rename extensions back to yaml, keeping templates as yml.j2

Commit:
[patch] Review some more extensions

Commit:
[patch] fixing eof new line (lint)

Commit:
[patch] eof new line - part 2

Commit:
[patch] Fix playbook extension

durera committed:
Fix typo

durera committed:
Fix wrong file extension

durera committed:
Tidy up tekton build

Implement standard structure, and integrate all tasks into the build
- src
- tmp
- target

durera committed:
Not creating the with-fvt version atm

durera committed:
Missed a file in the last commit

durera committed:
More updates from clitest env

durera committed:
Another yaml -> yml change that needs to be put back

durera committed:
Fix comment formatting
